contrib/nexusproto/serializer.go

```diff
@@ -0,0 +1,159 @@
+package nexusproto
+
+import (
+	"errors"
+	"fmt"
+	"mime"
+	"reflect"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+func messageFromAny(v any) (proto.Message, error) {
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Ptr {
+		return nil, fmt.Errorf("%w: type: %T", ErrInvalidProtoObject, v)
+	}
+	elem := rv.Type().Elem()
+
+	// v is a proto.Message or *proto.Message, the serializer supports either:
+	// 1. var d durationpb.Duration; s.Deserialize(&d)
+	// 2. var d *durationpb.Duration; s.Deserialize(&d)
+	// In the second case, we need to instantiate an empty proto struct:
+	if elem.Kind() == reflect.Ptr {
+		empty := reflect.New(elem.Elem())
+		rv.Elem().Set(empty)
+		rv = empty
+	}
+
+	msg, ok := rv.Interface().(proto.Message)
+	if !ok {
+		return nil, nexus.ErrSerializerIncompatible
```

Comment:
I see we are using the same error for serialize and deserialize, even though the error for serialize is meant to be a benign error and deserialize is meant to be a user-facing error, is that correct? I think we should use a more descriptive error here saying the content is a protobuf message but the type is not.

Reply:
For serialize it may not be benign if we exhaust all serializers. I will add more info to the message.